src/main/scala/org/constellation/gossip/GossipService.scala

```diff
@@ -1,47 +1,123 @@
 package org.constellation.gossip
 
-import cats.data.OptionT
-import cats.effect.Concurrent
-import cats.implicits._
+import cats.Parallel
+import cats.effect.{Concurrent, Timer}
+import cats.syntax.all._
+import io.chrisdavenport.log4cats.SelfAwareStructuredLogger
+import io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+import org.constellation.gossip.bisect.bisectA
 import org.constellation.gossip.sampling.PeerSampling
-import org.constellation.infrastructure.p2p.ClientInterpreter
+import org.constellation.gossip.state.{GossipMessage, GossipMessagePathTracker}
 import org.constellation.infrastructure.p2p.PeerResponse.PeerClientMetadata
 import org.constellation.p2p.Cluster
-import org.constellation.schema.{Id, NodeState}
+import org.constellation.schema.Id
 
-abstract class GossipService[F[_], A](selfId: Id, peerSampling: PeerSampling[F, Id], cluster: Cluster[F])(
-  implicit val F: Concurrent[F]
+import scala.concurrent.duration._
+
+abstract class GossipService[F[_]: Parallel, A](
+  selfId: Id,
+  peerSampling: PeerSampling[F],
+  cluster: Cluster[F],
+  messageTracker: GossipMessagePathTracker[F, A]
+)(
+  implicit F: Concurrent[F],
+  T: Timer[F]
 ) {
+  protected val logger: SelfAwareStructuredLogger[F] = Slf4jLogger.getLogger[F]
+
+  /**
+    * How to spread the message to peer
+    */
+  protected def spreadFn(peerClientMetadata: PeerClientMetadata, message: GossipMessage[A]): F[Unit]
+
+  /**
+    * How to check if message was received by peer
+    */
+  protected def validationFn(peerClientMetadata: PeerClientMetadata, message: GossipMessage[A]): F[Boolean]
+
+  def recover(message: GossipMessage[A]): F[Unit] = {
+    val sequence = message.path.toIndexedSeq
+
+    for {
+      failureNode <- bisectA(
+        (id: Id) => getClientMetadata(id).flatMap(validationFn(_, message))
+      )(sequence)
+      sequenceToRetry = sequence.dropWhile(failureNode.contains)
+      clients <- sequenceToRetry.toList.traverse(getClientMetadata)
+      _ <- clients.traverse(spreadFn(_, message))
+    } yield ()
+  }
 
-  protected def spread(data: A, fanout: Int)(spreadFn: (PeerClientMetadata, GossipMessage[A]) => F[Unit]): F[Unit] =
+  /**
+    * Start spreading the message to fanout
+    */
+  def spread(data: A, fanout: Int): F[Unit] =
     for {
+      _ <- logger.debug(s"Starting spreading of data: $data with fanout $fanout")
       messages <- peerSampling
         .selectPaths(fanout)
-        .map(_.map(GossipPath(_)).map(GossipMessage(data, _)))
+        .map(_.map(GossipMessage(data, _)))
+      _ <- messages.map { msg =>
+        for {
+          _ <- logger.debug(
+            s"Path ${msg.path.id}: ${msg.path.toIndexedSeq.map(_.short).mkString(" -> ")}"
+          )
+          _ <- messageTracker.start(msg)
+          _ <- spreadToNext(msg)
+          _ <- logger.debug(s"Started path ${msg.path.id}")
+          _ <- T.sleep(30 seconds)
+          succeeded <- messageTracker.isSuccess(msg)
```

Comment:
We should just use .onSuccess on the api calls in spreadToNext and update the message tracker.

Reply:
`isSuccess` says if whole path succeeded. Not only if sending to layer 1 succeeded.

Reply:
Basically we agreed that we just "fire and forget" the message to layer 1 and we just store the information that we sent the message to layer 1. Then we just wait for the message back from last layer or we just wait 30 seconds (or any timeout) and check if it managed to reach the last layer or failed somewhere.